server/src/main/java/org/elasticsearch/gateway/GatewayMetaState.java

```diff
@@ -54,135 +57,254 @@
-                try {
-                    metaStateService.writeGlobalState("changed", newMetaData);
-                } catch (Exception e) {
-                    success = false;
-                }
+            updateMetaData(event);
+        } catch (Exception e) {
+            logger.warn("Exception occurred when storing new meta data", e);
+        }
+    }
+
+    /**
+     * This class is used to write changed global {@link MetaData}, {@link IndexMetaData} and {@link Manifest} to disk.
+     * This class delegates <code>write*</code> calls to corresponding write calls in {@link MetaStateService} and
+     * additionally it keeps track of cleanup actions to be performed if transaction succeeds or fails.
+     * Transaction succeeds if {@link #writeManifest(String, Manifest)} call succeeds, transaction fails if
+     * any <code>write*</code> call fails.
+     * Once transaction succeeds/fails it can no longer be used.
+     */
+    static class Transaction {
+        private static final String TRANSACTION_COMPLETED_MSG = "Transaction is already completed";
+        private final List<Runnable> commitCleanupActions;
+        private final List<Runnable> rollbackCleanupActions;
+        private final Manifest previousManifest;
+        private final MetaStateService metaStateService;
+        private boolean finished;
+
+        Transaction(MetaStateService metaStateService, Manifest previousManifest) {
+            this.metaStateService = metaStateService;
+            assert previousManifest != null;
+            this.previousManifest = previousManifest;
+            this.commitCleanupActions = new ArrayList<>();
+            this.rollbackCleanupActions = new ArrayList<>();
+            this.finished = false;
+        }
+
+        long writeGlobalState(String reason, MetaData metaData) throws WriteStateException {
+            assert finished == false : TRANSACTION_COMPLETED_MSG;
+            try {
+                long generation = metaStateService.writeGlobalState(reason, metaData);
+                commitCleanupActions.add(() -> metaStateService.cleanupGlobalState(generation));
+                rollbackCleanupActions.add(() -> metaStateService.cleanupGlobalState(previousManifest.getGlobalGeneration()));
+                return generation;
+            } catch (WriteStateException e) {
+                rollback();
+                throw e;
             }
+        }
 
+        long writeIndex(String reason, IndexMetaData metaData) throws WriteStateException {
+            assert finished == false : TRANSACTION_COMPLETED_MSG;
+            try {
+                Index index = metaData.getIndex();
+                long generation = metaStateService.writeIndex(reason, metaData);
+                commitCleanupActions.add(() -> metaStateService.cleanupIndex(index, generation));
+                long previousGeneration = previousManifest.getIndexGenerations().getOrDefault(index, -1L);
```

Comment:
@ywelsch If the previous manifest file does not reference index metadata what is the point to keep this metadata around?

Reply:
- We want to typically keep the index metadata around as long as we keep the shard data around (so that we can reimport the data as dangling index).
- A node might have shard data but that shard data might not be allocated (e.g. because allocation failed).
- The shard deletion logic only starts kicking in once the shard is green (i.e. there are enough other shard copies around in the cluster).

Assume the scenario where you have a data node that has a shard allocated, then shard fails, and the shard is allocated to another node. The node will stop writing index metadata for that failed shard, but not delete the shard unless the other node has completed recovery. Assume that we now again want to allocate the previously failed shard to this node, but writing the index metadata fails, we will then be cleaning up the previous (successfully) written index metadata, effectively destroying on-disk information that could be useful for dangling indices. I know this is quite the edge case, and we can definitely debate on whether we want to address it.
I wonder if we should be extra careful here. One solution would be not to do this clean-up when having the -1 case. I know that that has other unwanted consequences (i.e. on repeated failures writing out the state files again and again). I would rather err on the side of not mistakenly destroying data though.

Reply:
Ok, https://github.com/elastic/elasticsearch/pull/35049/commits/0bc5ab55c7d6479839cafd43ed189d5bc1802306